Generate different outputs depending on the backend and update to adam-robotics 0.3.2

flferretti opened the pull request:
When this package is installed, it also installs some quite heavy dependencies that might not be needed, namely JAX and PyTorch. This PR allows to generate multiple outputs from this recipe depending on the chosen backend (`casadi`, `jax`, `pytorch`, `all`).

> [!Warning]
> With these changes, installing the plain `adam-robotics` will NOT install `casadi`, `jax` and `pytorch` backends. 
> Please, use `adam-robotics-all` to install all the backends, or install `adam-robotics-casadi` to install `adam-robotics` with the `casadi` backend.

<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->

Fix https://github.com/conda-forge/adam-robotics-feedstock/pull/13 .

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

I do have some suggestions for making it better though...


For **recipe/meta.yaml**:

 * ℹ️ It looks like the 'adam-robotics-all' output doesn't have any tests.

traversaro commented on recipe/meta.yaml:
```diff
@@ -10,46 +11,145 @@ source:
+
+  - name: {{ name|lower }}-all
+    build:
+      skip: True  # [win]
```
Could it make sense to also have `adam-robotics-all` also on Windows, and just install all the packages that are available there? Or it may be too confusing?

flferretti replied:
Yes, it makes sense. Thanks!

flferretti replied:
Done in https://github.com/conda-forge/adam-robotics-feedstock/pull/12/commits/611b8fc016d0ca01c49a0c730fbf825b3bc72b26

traversaro commented:
Just to understand, before when one installed `adam-robotics` from conda-forge, it would also get `casadi` installed as a dependency, while after this PR, no, is that correct? Probably we can document this clearly in the PR description.

flferretti commented:
> Just to understand, before when one installed `adam-robotics` from conda-forge, it would also get `casadi` installed as a dependency, while after this PR, no, is that correct? Probably we can document this clearly in the PR description.

That's correct! I'll update the description. Thanks!

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

flferretti commented:
Tests are failing because of `geometry` module that will be removed by https://github.com/ami-iit/adam/pull/106

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found some lint.

Here's what I've got...


For **recipe/meta.yaml**:

 * ❌ Recipe maintainer "xela-95" does not exist

flferretti commented:
Test are actually passing, but the feedstock does not have the permissions to produce different outputs than the one originally associated to its token, see https://conda-forge.org/docs/maintainer/infrastructure/#output-validation-and-feedstock-tokens. If this PR gets approved, I'll have to send a request according to https://github.com/conda-forge/admin-requests?tab=readme-ov-file#add-a-package-output-to-a-feedstock to enable additional outputs.

FYI @traversaro

flferretti commented:
Note that in https://github.com/conda-forge/adam-robotics-feedstock/pull/12/commits/c16f10b58deb341e25f54e1e8b3aad66c4da9e82 I added already a rebuild for the version 0.3.2 so we can easily rebase onto #13, and in https://github.com/conda-forge/adam-robotics-feedstock/pull/12/commits/34abe6ace164fdec6fc43e0adc50ff2d308c9582 I added a fix for the problem we were facing in #10. It was related to the fact that the MAX_PATH length is set to 260, which makes a check fail during the build. I added a patch to remove it, but we can eventually consider to remove also upstream

FYI @Giulero @xela-95

traversaro commented:
> I'll have to send a request according to https://github.com/conda-forge/admin-requests?tab=readme-ov-file#add-a-package-output-to-a-feedstock to enable additional outputs.

For me we can proceed, thanks!

traversaro commented:
Fyi pytorch is basically not available on Windows. However, I am not sure if multiplying the amount of build for 12 (and it should be even more, as we should also add `linux-aarch64` and `osx-arm64` ) is worth the minimal convenience of having `adam-robotics-all` installable also on Windows, where it should effectively just install `adam-robotics-casadi`. Perhaps we should make also `adam-robotics-all` a noarch package (that will not install in windows as the dependencies are missing) so that we can make the recipe a `noarch: python` package again?

traversaro commented:
By the way, there seems to be test failures on Python 3.9 : 

~~~
=========================== short test summary info ============================
ERROR tests/body_fixed/test_Jax_body_fixed.py - TypeError: unsupported operand type(s) for |: 'type' and 'type'
ERROR tests/body_fixed/test_NumPy_body_fixed.py - TypeError: unsupported operand type(s) for |: 'type' and 'type'
ERROR tests/mixed/test_Jax_mixed.py - TypeError: unsupported operand type(s) for |: 'type' and 'type'
ERROR tests/mixed/test_NumPy_mixed.py - TypeError: unsupported operand type(s) for |: 'type' and 'type'
ERROR tests/parametric/test_Jax_computations_parametric.py - TypeError: unsupported operand type(s) for |: 'type' and 'type'
ERROR tests/parametric/test_NumPy_computations_parametric.py - TypeError: unsupported operand type(s) for |: 'type' and 'type'
!!!!!!!!!!!!!!!!!!! Interrupted: 6 errors during collection !!!!!!!!!!!!!!!!!!!!
======================== 3 warnings, 6 errors in 1.23s =========================
~~~

are we sure we still support Python 3.9 here?

flferretti commented:
> By the way, there seems to be test failures on Python 3.9 :
> 
> ```
> =========================== short test summary info ============================
> ERROR tests/body_fixed/test_Jax_body_fixed.py - TypeError: unsupported operand type(s) for |: 'type' and 'type'
> ERROR tests/body_fixed/test_NumPy_body_fixed.py - TypeError: unsupported operand type(s) for |: 'type' and 'type'
> ERROR tests/mixed/test_Jax_mixed.py - TypeError: unsupported operand type(s) for |: 'type' and 'type'
> ERROR tests/mixed/test_NumPy_mixed.py - TypeError: unsupported operand type(s) for |: 'type' and 'type'
> ERROR tests/parametric/test_Jax_computations_parametric.py - TypeError: unsupported operand type(s) for |: 'type' and 'type'
> ERROR tests/parametric/test_NumPy_computations_parametric.py - TypeError: unsupported operand type(s) for |: 'type' and 'type'
> !!!!!!!!!!!!!!!!!!! Interrupted: 6 errors during collection !!!!!!!!!!!!!!!!!!!!
> ======================== 3 warnings, 6 errors in 1.23s =========================
> ```
> 
> are we sure we still support Python 3.9 here?

No, using pipe for union types has been introduced in Python 3.10 according to [PEP 604](https://peps.python.org/pep-0604/). This should be changed in the package's `setup.cfg`. For the time being, I'll put a `{{ python_min }}` bound here

flferretti commented:
> Fyi pytorch is basically not available on Windows. However, I am not sure if multiplying the amount of build for 12 (and it should be even more, as we should also add `linux-aarch64` and `osx-arm64` ) is worth the minimal convenience of having `adam-robotics-all` installable also on Windows, where it should effectively just install `adam-robotics-casadi`. Perhaps we should make also `adam-robotics-all` a noarch package (that will not install in windows as the dependencies are missing) so that we can make the recipe a `noarch: python` package again?

You're right, thanks! I restored the `noarch` and set minimum python version in https://github.com/conda-forge/adam-robotics-feedstock/pull/12/commits/766d406f2b24f176911490932e1ebba38c72eafd

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

traversaro commented:
The run constraint should be `python >={{ python_min }}`, not `python {{ python_min }}` (that constraints the run python version to be the minimum one). Furthermore, if we want to override the minimum version with ours that is different (i.e. 3.10), we should add a line `{% set python_min = "3.10" %}` at the top of the `meta.yaml` . For more info, see https://conda-forge.org/docs/maintainer/knowledge_base/#noarch-python .

flferretti commented:
> > I'll have to send a request according to https://github.com/conda-forge/admin-requests?tab=readme-ov-file#add-a-package-output-to-a-feedstock to enable additional outputs.
> 
> For me we can proceed, thanks!

Done in https://github.com/conda-forge/admin-requests/pull/1172, thanks!

flferretti commented:
> override the minimum version with ours that is different

You're right, sorry. Should be solved by rebase in https://github.com/conda-forge/adam-robotics-feedstock/pull/12/commits/4510fcbc6c2a2d8c8c4e96f6092adacac1a31952

flferretti commented:
For some reason I still see `.ci_support/linux_64_numpy1.22python3.9.____cpython.yaml` in the PR files, @traversaro am I doing something wrong?

traversaro commented:
@conda-forge-admin please rerender

traversaro commented:
> For some reason I still see `.ci_support/linux_64_numpy1.22python3.9.____cpython.yaml` in the PR files, @traversaro am I doing something wrong?

I am not sure, but are you sure that we need a python build dependencies in https://github.com/conda-forge/adam-robotics-feedstock/pull/12/files#diff-f3725a55bf339595bf865fec73bda8ac99f283b0810c205442021f29c06eea9aR31 ? According to https://conda-forge.org/docs/maintainer/knowledge_base/#noarch-python this should not be necessary.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-webservice.

I tried to rerender for you, but it looks like there was nothing to do.


<sub>This message was generated by GitHub actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/11913720545](https://github.com/conda-forge/conda-forge-webservices/actions/runs/11913720545).</sub>

traversaro commented:
Probably you can also try marking the whole recipe as `noarch: python` on the top (or instead, not sure) of each output.

traversaro commented on recipe/meta.yaml:
```diff
@@ -1,54 +1,168 @@
+      source_files:
+        - tests
+      commands:
+        - pytest ./tests/ -v -k "not gravity"  # [not win]
+        - pytest ./tests/ -v -k "not gravity" --ignore-glob=*Jax*  # [win]
+      requires:
+        - python  >={{ python_min }}
+        - pip
+        - pytest
+        - icub-models
+        - gym-ignition-models
+        - idyntree
+        - git
+        - gitpython
+        - casadi
+        - pytorch
+        - jax  # [not win]
```
To avoid confusing the bot, perhaps we can remove the `# [win]` selectors, as anyhow the `-all` package at this point is not installable on Windows.
```suggestion
        - pytest ./tests/ -v -k "not gravity"
      requires:
        - python  >={{ python_min }}
        - pip
        - pytest
        - icub-models
        - gym-ignition-models
        - idyntree
        - git
        - gitpython
        - casadi
        - pytorch
        - jax
```

flferretti replied:
Thanks! Addressed in https://github.com/conda-forge/adam-robotics-feedstock/pull/12/commits/59c845cd57a37d4bc6d8bfb3371bb1529c05c3a5 with rebase

traversaro commented on recipe/meta.yaml:
```diff
@@ -1,54 +1,168 @@
+      commands:
+        - pytest ./tests/ -v -k "not gravity" --ignore-glob=*Jax* --ignore-glob=*CasADi* --ignore-glob=*idyntree*
+      requires:
+        - python  >={{ python_min }}
```
```suggestion
        - python  {{ python_min }}
```

See https://conda-forge.org/docs/maintainer/knowledge_base/#noarch-python for more details. The idea is that you want to check in tests if the minimum supported version is actually working (so you want to pin it) while for run any version newer is ok.

flferretti replied:
Thanks! Addressed in https://github.com/conda-forge/adam-robotics-feedstock/pull/12/commits/59c845cd57a37d4bc6d8bfb3371bb1529c05c3a5 with rebase

flferretti commented:
Probably `adam-robotics-pytorch` needs a newer python version as the recipe gets solved with `jax==0.3.0` and this leads to:

```python
__________ ERROR collecting tests/pytorch_batch/test_pytorch_batch.py __________
tests/pytorch_batch/test_pytorch_batch.py:5: in <module>
    import jax.numpy as jnp
../_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/python3.10/site-packages/jax/__init__.py:133: in <module>
    from jax import nn as nn
../_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/python3.10/site-packages/jax/nn/__init__.py:21: in <module>
    from jax._src.nn.functions import (
../_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/python3.10/site-packages/jax/_src/nn/functions.py:28: in <module>
    from jax.scipy.special import expit
../_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/python3.10/site-packages/jax/scipy/__init__.py:17: in <module>
    from jax.scipy import linalg as linalg
../_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/python3.10/site-packages/jax/scipy/linalg.py:17: in <module>
    from jax._src.scipy.linalg import (
../_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/python3.10/site-packages/jax/_src/scipy/linalg.py:242: in <module>
    @_wraps(scipy.linalg.tril)
E   AttributeError: module 'scipy.linalg' has no attribute 'tril'
```

<details><summary>Test Environment</summary>
<p>

```yaml

The following NEW packages will be INSTALLED:

    _libgcc_mutex:         0.1-conda_forge                 conda-forge
    _openmp_mutex:         4.5-2_kmp_llvm                  conda-forge
    absl-py:               2.1.0-pyhd8ed1ab_0              conda-forge
    adam-robotics:         0.3.2-pyhe714708_0              local      
    adam-robotics-pytorch: 0.3.2-pyhee61e44_0              local      
    ampl-mp:               3.1.0-h2cc385e_1006             conda-forge
    assimp:                5.4.3-h8943939_0                conda-forge
    attr:                  2.5.1-h166bdaf_1                conda-forge
    bzip2:                 1.0.8-h4bc722e_7                conda-forge
    c-ares:                1.32.3-h4bc722e_0               conda-forge
    ca-certificates:       2024.8.30-hbcca054_0            conda-forge
    colorama:              0.4.6-pyhd8ed1ab_0              conda-forge
    dbus:                  1.13.6-h5008d03_3               conda-forge
    eigen:                 3.4.0-h00ab1b0_0                conda-forge
    exceptiongroup:        1.2.2-pyhd8ed1ab_0              conda-forge
    expat:                 2.6.4-h5888daf_0                conda-forge
    filelock:              3.16.1-pyhd8ed1ab_0             conda-forge
    fsspec:                2024.10.0-pyhff2d567_0          conda-forge
    gettext:               0.22.5-he02047a_3               conda-forge
    gettext-tools:         0.22.5-he02047a_3               conda-forge
    git:                   2.47.0-pl5321h59d505e_0         conda-forge
    gitdb:                 4.0.11-pyhd8ed1ab_0             conda-forge
    gitpython:             3.1.43-pyhd8ed1ab_0             conda-forge
    glfw:                  3.4-hd590300_0                  conda-forge
    gmp:                   6.3.0-hac33072_2                conda-forge
    gmpy2:                 2.1.5-py310he8512ff_2           conda-forge
    gym-ignition-models:   1.1.1-pyhd8ed1ab_0              conda-forge
    icu:                   75.1-he02047a_0                 conda-forge
    icub-models:           2.8.1-py310h7545b00_0           conda-forge
    idyntree:              13.1.1-py310h2e4eca6_0          conda-forge
    iniconfig:             2.0.0-pyhd8ed1ab_0              conda-forge
    ipopt:                 3.14.16-h122424a_10             conda-forge
    irrlicht:              1.8.5-hcce6d95_5                conda-forge
    jax:                   0.3.0-pyhd8ed1ab_0              conda-forge
    jax2torch:             0.0.7-pyhd8ed1ab_0              conda-forge
    jaxlib:                0.1.75-py310hb5077e9_0          conda-forge
    jinja2:                3.1.4-pyhd8ed1ab_0              conda-forge
    keyutils:              1.6.1-h166bdaf_0                conda-forge
    krb5:                  1.21.3-h659f571_0               conda-forge
    lame:                  3.100-h166bdaf_1003             conda-forge
    ld_impl_linux-64:      2.43-h712a8e2_2                 conda-forge
    libabseil:             20240722.0-cxx17_h5888daf_1     conda-forge
    libasprintf:           0.22.5-he8f35ee_3               conda-forge
    libasprintf-devel:     0.22.5-he8f35ee_3               conda-forge
    libblas:               3.9.0-25_linux64_openblas       conda-forge
    libboost:              1.86.0-hb8260a3_2               conda-forge
    libcap:                2.69-h0f662aa_0                 conda-forge
    libcblas:              3.9.0-25_linux64_openblas       conda-forge
    libcurl:               8.10.1-hbbe4b11_0               conda-forge
    libdrm:                2.4.123-hb9d3cd8_0              conda-forge
    libedit:               3.1.20191231-he28a2e2_2         conda-forge
    libegl:                1.7.0-ha4b6fd6_2                conda-forge
    libev:                 4.33-hd590300_2                 conda-forge
    libexpat:              2.6.4-h5888daf_0                conda-forge
    libffi:                3.4.2-h7f98852_5                conda-forge
    libflac:               1.4.3-h59595ed_0                conda-forge
    libgcc:                14.2.0-h77fa898_1               conda-forge
    libgcc-ng:             14.2.0-h69a702a_1               conda-forge
    libgcrypt:             1.11.0-h4ab18f5_1               conda-forge
    libgettextpo:          0.22.5-he02047a_3               conda-forge
    libgettextpo-devel:    0.22.5-he02047a_3               conda-forge
    libgfortran:           14.2.0-h69a702a_1               conda-forge
    libgfortran-ng:        14.2.0-h69a702a_1               conda-forge
    libgfortran5:          14.2.0-hd5240d6_1               conda-forge
    libgl:                 1.7.0-ha4b6fd6_2                conda-forge
    libglib:               2.82.2-h2ff4ddf_0               conda-forge
    libglu:                9.0.3-h03adeef_0                conda-forge
    libglvnd:              1.7.0-ha4b6fd6_2                conda-forge
    libglx:                1.7.0-ha4b6fd6_2                conda-forge
    libgpg-error:          1.51-hbd13f7d_1                 conda-forge
    libhwloc:              2.11.1-default_hecaa2ac_1000    conda-forge
    libiconv:              1.17-hd590300_2                 conda-forge
    libjpeg-turbo:         3.0.0-hd590300_1                conda-forge
    liblapack:             3.9.0-25_linux64_openblas       conda-forge
    libnghttp2:            1.64.0-h161d5f1_0               conda-forge
    libnsl:                2.0.1-hd590300_0                conda-forge
    libogg:                1.3.5-h4ab18f5_0                conda-forge
    libopenblas:           0.3.28-pthreads_h94d23a6_1      conda-forge
    libopus:               1.3.1-h7f98852_1                conda-forge
    libosqp:               0.6.3-h5888daf_1                conda-forge
    libpciaccess:          0.18-hd590300_0                 conda-forge
    libpng:                1.6.44-hadc24fc_0               conda-forge
    libprotobuf:           5.28.2-h5b01275_0               conda-forge
    libqdldl:              0.1.7-hcb278e6_0                conda-forge
    libscotch:             7.0.4-h2fe6a88_5                conda-forge
    libsndfile:            1.2.2-hc60ed4a_1                conda-forge
    libspral:              2024.05.08-h831f25b_3           conda-forge
    libsqlite:             3.47.0-hadc24fc_1               conda-forge
    libssh2:               1.11.0-h0841786_0               conda-forge
    libstdcxx:             14.2.0-hc0a3c3a_1               conda-forge
    libstdcxx-ng:          14.2.0-h4852527_1               conda-forge
    libsystemd0:           256.7-h2774228_1                conda-forge
    libtorch:              2.5.1-cpu_mkl_hb9d73ce_103      conda-forge
    libuuid:               2.38.1-h0b41bf4_0               conda-forge
    libuv:                 1.49.2-hb9d3cd8_0               conda-forge
    libvorbis:             1.3.7-h9c3ff4c_0                conda-forge
    libxcb:                1.17.0-h8a09558_0               conda-forge
    libxcrypt:             4.4.36-hd590300_1               conda-forge
    libxkbcommon:          1.7.0-h2c5496b_1                conda-forge
    libxml2:               2.13.5-hb346dea_0               conda-forge
    libxslt:               1.1.39-h76b75d6_0               conda-forge
    libzlib:               1.3.1-hb9d3cd8_2                conda-forge
    llvm-openmp:           19.1.3-h024ca30_0               conda-forge
    lxml:                  5.3.0-py310h6ee67d5_2           conda-forge
    lz4-c:                 1.9.4-hcb278e6_0                conda-forge
    markupsafe:            3.0.2-py310h89163eb_0           conda-forge
    metis:                 5.1.0-hd0bcaf9_1007             conda-forge
    mkl:                   2024.2.2-ha957f24_16            conda-forge
    mpc:                   1.3.1-h24ddda3_1                conda-forge
    mpfr:                  4.2.1-h90cbb55_3                conda-forge
    mpg123:                1.32.9-hc50e24c_0               conda-forge
    mpmath:                1.3.0-pyhd8ed1ab_0              conda-forge
    mumps-include:         5.7.3-ha770c72_5                conda-forge
    mumps-seq:             5.7.3-h27a6a8b_0                conda-forge
    ncurses:               6.5-he02047a_1                  conda-forge
    networkx:              3.4.2-pyh267e887_2              conda-forge
    numpy:                 1.26.4-py310hb13e2d6_0          conda-forge
    openssl:               3.4.0-hb9d3cd8_0                conda-forge
    opt_einsum:            3.4.0-pyhd8ed1ab_0              conda-forge
    osqp-eigen:            0.8.1-hdd734ac_1                conda-forge
    packaging:             24.2-pyhff2d567_1               conda-forge
    pcre2:                 10.44-hba22ea6_2                conda-forge
    perl:                  5.32.1-7_hd590300_perl5         conda-forge
    pip:                   24.3.1-pyh8b19718_0             conda-forge
    pluggy:                1.5.0-pyhd8ed1ab_0              conda-forge
    prettytable:           3.12.0-pyhd8ed1ab_0             conda-forge
    pthread-stubs:         0.4-hb9d3cd8_1002               conda-forge
    pulseaudio-client:     17.0-hb77b528_0                 conda-forge
    pytest:                8.3.3-pyhd8ed1ab_0              conda-forge
    python:                3.10.0-h543edf9_3_cpython       conda-forge
    python-flatbuffers:    24.3.25-pyh59ac667_0            conda-forge
    python_abi:            3.10-5_cp310                    conda-forge
    pytorch:               2.5.1-cpu_mkl_py310h89e431c_103 conda-forge
    pyyaml:                6.0.2-py310ha75aee5_1           conda-forge
    readline:              8.2-h8228510_1                  conda-forge
    scipy:                 1.14.1-py310hfcf56fc_1          conda-forge
    sdl:                   1.2.68-h293081c_0               conda-forge
    sdl2:                  2.30.7-h3ed165c_0               conda-forge
    setuptools:            75.5.0-pyhff2d567_0             conda-forge
    six:                   1.16.0-pyh6c4a22f_0             conda-forge
    sleef:                 3.7-h1b44611_2                  conda-forge
    smmap:                 5.0.0-pyhd8ed1ab_0              conda-forge
    sqlite:                3.47.0-h9eae976_1               conda-forge
    sympy:                 1.13.3-pypyh2585a3b_103         conda-forge
    tbb:                   2021.13.0-h84d6215_0            conda-forge
    tk:                    8.6.13-noxft_h4845f30_101       conda-forge
    tomli:                 2.1.0-pyhff2d567_0              conda-forge
    typing_extensions:     4.12.2-pyha770c72_0             conda-forge
    tzdata:                2024b-hc8b5060_0                conda-forge
    unixodbc:              2.3.12-h661eb56_0               conda-forge
    urdfdom-py:            1.2.1-py310hff52083_5           conda-forge
    wayland:               1.23.1-h3e06ad9_0               conda-forge
    wcwidth:               0.2.13-pyhd8ed1ab_0             conda-forge
    wheel:                 0.45.0-pyhd8ed1ab_0             conda-forge
    xcb-util:              0.4.1-hb711507_2                conda-forge
    xkeyboard-config:      2.43-hb9d3cd8_0                 conda-forge
    xorg-libx11:           1.8.10-h4f16b4b_0               conda-forge
    xorg-libxau:           1.0.11-hb9d3cd8_1               conda-forge
    xorg-libxdamage:       1.1.6-hb9d3cd8_0                conda-forge
    xorg-libxdmcp:         1.1.5-hb9d3cd8_0                conda-forge
    xorg-libxext:          1.3.6-hb9d3cd8_0                conda-forge
    xorg-libxfixes:        6.0.1-hb9d3cd8_0                conda-forge
    xorg-libxinerama:      1.1.5-h5888daf_1                conda-forge
    xorg-libxrandr:        1.5.4-hb9d3cd8_0                conda-forge
    xorg-libxrender:       0.9.11-hb9d3cd8_1               conda-forge
    xorg-libxxf86vm:       1.1.5-hb9d3cd8_4                conda-forge
    xorg-xorgproto:        2024.1-hb9d3cd8_1               conda-forge
    xz:                    5.2.6-h166bdaf_0                conda-forge
    yaml:                  0.2.5-h7f98852_2                conda-forge
    zlib:                  1.3.1-hb9d3cd8_2                conda-forge
    zstd:                  1.5.6-ha6fb4c9_0                conda-forge
```

</p>
</details> 

`scipy.linalg.tril` has been deprecated in `scipy==1.11.0`, see https://docs.scipy.org/doc/scipy-1.12.0/reference/generated/scipy.linalg.tril.html

flferretti commented:
When forcing `jax>0.3.0` (in red) I get this diff in the resolved environment. It seems like it has to do with `python` and `numpy`:
```diff
71c73
<   jax                conda-forge/noarch::jax-0.4.34-pyhd8ed1ab_0
---
>   jax                conda-forge/noarch::jax-0.3.0-pyhd8ed1ab_0
73c75
<   jaxlib             conda-forge/linux-64::jaxlib-0.4.34-cuda120py313h3b1fb80_200
---
>   jaxlib             conda-forge/linux-64::jaxlib-0.1.75-py310hb5077e9_0
140c140
<   libtorch           conda-forge/linux-64::libtorch-2.4.1-cuda120_h1d34654_302
---
>   libtorch           conda-forge/linux-64::libtorch-2.5.1-cuda126_hd8d7199_303
153,154c153
<   mkl                conda-forge/linux-64::mkl-2023.2.0-h84fe81f_50496
<   ml_dtypes          conda-forge/linux-64::ml_dtypes-0.5.0-py313ha87cce1_0
---
>   mkl                conda-forge/linux-64::mkl-2024.2.2-ha957f24_16
164c163
<   numpy              conda-forge/linux-64::numpy-2.1.3-py313h4bf6692_0
---
>   numpy              conda-forge/linux-64::numpy-1.26.4-py310hb13e2d6_0
174,177c172,175
<   python             conda-forge/linux-64::python-3.13.0-h9ebbce0_100_cp313
<   python_abi         conda-forge/linux-64::python_abi-3.13-5_cp313
<   pytorch            conda-forge/linux-64::pytorch-2.4.1-cuda120_py313h6ccb88c_302
<   re2                conda-forge/linux-64::re2-2024.07.02-h77b4e00_1
---
>   python             conda-forge/linux-64::python-3.10.15-h4a871b0_2_cpython
>   python-flatbuffers conda-forge/noarch::python-flatbuffers-24.3.25-pyh59ac667_0
>   python_abi         conda-forge/linux-64::python_abi-3.10-5_cp310
>   pytorch            conda-forge/linux-64::pytorch-2.5.1-cuda126_py310he4c8055_303
179c177
<   scipy              conda-forge/linux-64::scipy-1.14.1-py313h27c5614_1
---
>   scipy              conda-forge/linux-64::scipy-1.14.1-py310hfcf56fc_1
```

traversaro commented:
Trying to solve:

~~~
[project]
authors = ["Silvio <silvio.traversaro@iit.it>"]
channels = ["conda-forge"]
description = "Add a short description here"
name = "pixiws"
platforms = ["linux-64"]
version = "0.1.0"

[tasks]

[dependencies]
jax = ">=0.4.0"
numpy = ">=2.0.0"
pytorch = ">=2.5.0"
~~~

result in a bunch of protobuf errors, I guess it is related to https://github.com/conda-forge/jaxlib-feedstock/pull/287 and similar.

traversaro commented:
Probably we can just constrain `jax>=0.3.0` in the test session?

traversaro reviewed (approved):
Thanks a lot! Ready to go for me!

xela-95 reviewed (approved):
Thanks a lot @flferretti ! Impressive work!

Giulero reviewed (approved):
Awesome guys! Thank you for all the effort <3

traversaro commented:
Thanks to all!
